### ISSUE TITLE
EventMarkers Development

### ISSUE THREAD

**Issue by WinstonHChou** [jsimpso81/PathPlannerLabVIEW#9]
Hi @jsimpso81,
Is it possible for you to develop Event Markers execute functions for this library?
I've found it truely crucial for this year competition's auto.

**Comment by jsimpso81** [comment]
While not tested very well yet eventually marker triggering is in the newest version.   ( as is path finding using the Labview adstar ) I need to build the wrappers for pathfinding, update menus, create samples and publish.   Hooting to get that done today.  On Feb 3, 2024, at 5:17 AM, Winston ***@***.***> wrote:﻿
Hi @jsimpso81,
Is it possible for you to develop Event Markers execute functions for this library?
I've found it truely crucial for this year competition's auto.

**Comment by jsimpso81** [comment]
See the newest release.  It contains path following commands.   These include triggers.   Note that only boolean commands can be triggered.

**Comment by WinstonHChou** [comment]
@jsimpso81 yeah, i saw and it's wonderful you solve the boolean commands part.
I have developed the wait command and several functions on my own, but wasn't able to solve the boolean commands.
Would your want to take a look at my VIs?

edit:
I'm not sure if it's a good thing to do. But according to the characteristics of Race, Deadline Group, I decided to add a Timer per comand under an event.

**Comment by jsimpso81** [comment]
Yes.  If you can send them or send a link to your code

**Comment by WinstonHChou** [comment]
I would like to know if there's any way to detect a named-command has ended.
I was thinking to send a boolean command outside of the event execution part.
What's your thought?

Here's the Link: 
https://drive.google.com/drive/folders/1zGbDk8vqpjfaT_O3gPeOnT4w7AE9vHi3?usp=sharing

PathPlanner_CommandExecute.vi is what i expect to execute individual command
PathPlanner_EventMarkerExecute.vi (still testing) is what i expect to execute individual event

**Comment by jsimpso81** [comment]
As far as I can tell the path/trajectory continues without pauses until it ends.  Unless the speed is set to 0 the trajectory must continue to proceed.  The commands are intended to execute larger sequences of actions     For example. ShootThis command might - start shooting motors- wait until they were within correct speed - push note into shooter wheels - stop shooter motors- wait for next command.  So the single command shoot does a lot of little actions but these are not. ISI le to the path following.The wait for shooting motors would have a timeout to ensure there aren’t any endless loops.  The path would have to factor in the estimated maximum shooting timeDon’t know if this helps.  Hopefully it at least explains why wait wasn’t done.  Regardless I’ll try and incorporate wait if you let me see the code you did.  On Feb 4, 2024, at 5:33 PM, Winston ***@***.***> wrote:﻿
I would like to know if there's any way to detect a named-command has ended.
I was thinking to send a boolean command outside of the event execution part.
What's your thought?

**Comment by WinstonHChou** [comment]
I edited my last reply to include a google drvie link
You should be able to see the code from there.

**Comment by jsimpso81** [comment]
Thank you. On Feb 4, 2024, at 8:20 PM, Winston ***@***.***> wrote:﻿
I edited my last reply to include a google drvie link
You should be able to see the code from there.

**Comment by WinstonHChou** [comment]
my thought is to send a boolean signal back to the PathPlanner_CommandExecute.vi, and tells it a certain named-command ends.
But im struggle with both sending it from outside, and reading inside the CommandExecute.vi .

**Comment by WinstonHChou** [comment]
I feel like that the way you developed sending Boolean cmd inside this library would have a memory issue. Since whenever I'm test the function inside a while loop, when I stop the while loop, labview shows seriously delay.

**Comment by jsimpso81** [comment]
I’ll take a look. On Feb 6, 2024, at 4:48 AM, Winston ***@***.***> wrote:﻿
I feel like that the way you developed sending Boolean cmd inside this library would have a memory issue. Since whenever I'm test the function inside a while loop, when I stop the while loop, labview shows seriously delay.

**Comment by jsimpso81** [comment]
If you have a suggested fix send it. On Feb 6, 2024, at 4:48 AM, Winston ***@***.***> wrote:﻿
I feel like that the way you developed sending Boolean cmd inside this library would have a memory issue. Since whenever I'm test the function inside a while loop, when I stop the while loop, labview shows seriously delay.

**Comment by WinstonHChou** [comment]
I update a new version of my VIs. It works the way i want, but im not sure if they are working in the way you'd like to develope.
https://drive.google.com/drive/folders/1zGbDk8vqpjfaT_O3gPeOnT4w7AE9vHi3?usp=sharing

**Comment by WinstonHChou** [comment]
I found that Boolean Cmd Send and Receive would have a sampling frequency problem. I developed a boolean sampling filter with percentage pass (how many true there are in a sampling queue), but it doesn't help much. Is there another possible solution?

**Comment by WinstonHChou** [comment]
I should explain more on my thoughts of EventMarkers. I'd like to put every event and command that follows with the event inside of periodic task.vi, making it reusable for any other event and also different autonomous route(case).

**Comment by jsimpso81** [comment]
I finally got the issues with examples fixed.  I should be able to look at the code you provided later today.  I agree with your architecture.   I would suggest creating a number of commands that are sequences of actions.  Put the code for this in periodic tasks.   Trigger each one with a named boolean command.   Things like timers would go into these sequences.   For example:  command 1) Aim-and-shoot: actions - move arm to aim, when done (or timedout), start shooting motors, when up to speed (or timed out) push not into shooting wheels, wait a shooting time, stop motors.  Then wait for another command.  2) Pickup-from-floor: actions - move arm to floor, when in position (or timedout) run intake motor, wait until not is sensed (or timeout), stop intake motor.  Then wait for another command.  The WPILIB LabVIEW math library has a command multiplexor and drum controller that can be used to help create these commands.  They can also be created separately.    The commands could be triggered from autonomous or from teleop.   In autonomous a path could trigger the commands.  Also sequences of paths and command could be executed in sequence.  For example.  Aim-and-shoot command; check for done; run path; during path trigger intake; when path done, aim-and-shoot...   I'll probably get a better understanding of what you are thinking when I see your code.   

By the way, there is a new release of this code and a new release of the examples -- including a couple path finding examples.

**Comment by WinstonHChou** [comment]
Definitely gonna look into those suggestion.
Let me post some screenshots of our team's code for this part
![image](https://github.com/jsimpso81/PathPlannerLabVIEW/assets/68310078/203c0dd6-ca34-433a-9710-bf67aff0824b)
![image](https://github.com/jsimpso81/PathPlannerLabVIEW/assets/68310078/6de32879-9d62-40d2-8114-521cd64ce33f)
![image](https://github.com/jsimpso81/PathPlannerLabVIEW/assets/68310078/8061ea9c-64e0-42cf-a19b-23469e7d381e)

![image](https://github.com/jsimpso81/PathPlannerLabVIEW/assets/68310078/cc1dae0b-37fc-4b74-83f9-3c3faa8ee9c9)

![image](https://github.com/jsimpso81/PathPlannerLabVIEW/assets/68310078/4cffa81e-346e-4b8a-8012-ca6b61162e37)

![image](https://github.com/jsimpso81/PathPlannerLabVIEW/assets/68310078/29726576-2f83-404f-be33-f60e855ed7dd)

**Comment by WinstonHChou** [comment]
Issues that I encountered the most is about the memory usage for each vi. Since with preallocated VIs, a memory space matches an instance of that VI, which would run into issues with using same Feedback Node (or Shift Register) when executing different Command.

**Comment by WinstonHChou** [comment]
any news?
